staging/vhost-device-console/README.md

```diff
@@ -0,0 +1,133 @@
+  Note: The nested backend is selected by default and can be used only when
+        socket_count equals 1.
+
+## Limitations
+
+This device is still work-in-progress (WIP). The current version has been tested
+with VIRTIO_CONSOLE_F_MULTIPORT, but only for one console.
+
+## Features
+
+The current device gives access to multiple QEMU guest by providing a login prompt
+either by connecting to a localhost server port (network backend) or by creating an
+nested command prompt in the current terminal (nested backend). This prompt appears
+as soon as the guest is fully booted and gives the ability to user run command as a
+in regular terminal.
+
+## Examples
+
+### Dependencies
+For testing the device the required dependencies are:
+- Linux:
+    - Set `CONFIG_VIRTIO_CONSOLE=y`
+- QEMU (optional):
+    - A new vhost-user-console device has been implemented in the following repo:
+      - https://github.com/virtualopensystems/qemu/tree/vhu-console-rfc
+
+
+### Test the device
+
+The daemon should be started first:
+```shell
+host# vhost-device-console --socket-path=/tmp/console.sock --socket-count=1 \
+                           --tcp-port=12345 --backend=network
+```
+>Note: In case the backend is "nested" there is no need to provide
+       "--socket-count" and "--tcp-port" parameters.
+
+The QEMU invocation needs to create a chardev socket the device can
+use to communicate as well as share the guests memory over a memfd.
+
+There are two option for running QEMU with vhost-device-console:
+
+1) Using `vhost-user-console-pci`:
+```text
+host# qemu-system                                               \
+    <normal QEMU options>                                       \
+    -machine <machine options>,memory-backend=mem0              \
+    -object memory-backend-memfd,id=mem0,size=<Guest RAM size>  \ # size == -m size
+    -chardev socket,path=/tmp/console.sock0,id=con              \
+    -device vhost-user-console-pci,chardev=con0,id=console      \
+    ...
+```
+
+> Note: For testing this scenario the reader needs to clone the QEMU version from the following repo
+>       which implements `vhost-user-console` device.
+> - https://github.com/virtualopensystems/qemu/tree/vhu-console-rfc
+
+2) Using `vhost-user-device-pci`:
+```text
+host# qemu-system                                                                   \
+    <normal QEMU options>                                                           \
+    -machine <machine options>,memory-backend=mem0                                  \
+    -object memory-backend-memfd,id=mem0,size=<Guest RAM size>                      \ # size == -m size
+    -chardev socket,id=con0,path=/tmp/console.sock0                                 \
+    -device vhost-user-device-pci,chardev=con0,virtio-id=3,num_vqs=4,config_size=12 \
+    ...
+```
+
+Eventually, the user can connect to the console by running:
+```test
+host# stty raw && nc 127.0.0.1 12345
```

Comment:
I tested using this cmd line and I observed the following behavior:
```
root@syzkaller:~# llss^M
Connecting  Streaming    file.wav   nada.wav   rec.wav  test.wav
Paused      Unconnected  file2.wav  queen.wav  test.sh
root@syzkaller:~# 
```
Keys are displayed two times. Maybe my configuration is wrong. The command is correctly executed so guest is not getting the key two times.

Reply:
Ouch.. my bad. I have finally got it working with `stty -icanon -echo && nc localhost 12345 && stty echo` and also updated the README accordingly. Please let me know if that solves the issue on your side too.